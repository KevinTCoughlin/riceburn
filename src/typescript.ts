import * as ts from 'typescript';
import * as fs from 'fs';
import * as path from 'path';

function getSourceFile(file: string, content: string) {
  const normalizedFile = path.normalize(file).replace(/\\/g, '/');
  return ts.createSourceFile(normalizedFile, content, ts.ScriptTarget.ES5, true);
}

interface TypescriptMod {
  node: ts.Node;
  start: number;
  length: number;
  replacement: string;
}

const Modder = {
  replace(node: ts.Node, replacement: string): TypescriptMod {
    return {
      node,
      start: node.getStart(),
      length: node.getText().length,
      replacement
    };
  },

  append(node: ts.Node, suffix: string): TypescriptMod {
    return {
      node,
      start: node.getStart() + node.getText().length,
      length: 0,
      replacement: suffix
    };
  },

  prepend(node: ts.Node, prefix: string): TypescriptMod {
    return {
      node,
      start: node.getStart(),
      length: 0,
      replacement: prefix
    };
  },

  removeFull(node: ts.Node): TypescriptMod {
    return {
      node,
      start: node.getFullStart(),
      length: node.getFullText().length,
      replacement: ''
    };
  },

  remove(node: ts.Node): TypescriptMod {
    return {
      node,
      start: node.getStart(),
      length: node.getText().length,
      replacement: ''
    };
  }
};

function applyMods(content: string, mods: TypescriptMod[]) {
  let index = 0;
  let newContent: string[] = [];

  mods.forEach(mod => {
    newContent.push(content.slice(index, mod.start));
    newContent.push(mod.replacement);
    index = mod.start + mod.length;
  });

  newContent.push(content.slice(index));

  return newContent.join('');
}

function sortMods(mods: TypescriptMod[]) {
  const sorted = mods.sort((a, b) => {
    return a.start < b.start ? -1 : a.start === b.start ? 0 : 1;
  });

  // Check for overlapped ranges
  for (let i = 1; i < sorted.length; i++) {
    const prev = sorted[i - 1];
    const curr = sorted[i];
    if (prev.length + prev.start > curr.start) {
      throw new Error('Some of the mods overlap with each other. Please separate this out as another "asTypescript()" call.');
    }
  }

  return sorted;
}

export interface Visitor {
  // TODO: returns a TypescriptMod?
  (node: ts.Node, modder: typeof Modder): TypescriptMod | undefined;
}

export async function tsHandler(matches: string[], visitor: Visitor) {
  let sourceFile: ts.SourceFile;
  let content: string;
  let newContent: string;

  matches.forEach(async match => {
    content = fs.readFileSync(match).toString();

    try {
      sourceFile = getSourceFile(match, content);
    } catch {
      console.error('invalid Typescript file');
    }

    if (visitor) {
      let mods: TypescriptMod[] = [];

      const visitorWrapper = (node: ts.Node) => {
        const mod = visitor(node, Modder);
        if (mod) {
          mods.push(mod);
        }
        node.forEachChild(visitorWrapper);
      };

      sourceFile.forEachChild(visitorWrapper);

      newContent = applyMods(sourceFile.getFullText(), sortMods(mods));

      if (newContent !== content) {
        fs.writeFileSync(match, newContent);
      }
    }
  });
}
