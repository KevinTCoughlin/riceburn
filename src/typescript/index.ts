import ts from 'typescript';
import fs from 'fs';
import path from 'path';
import { modder } from './modder';
import { TypescriptMod, Visitor } from '../interfaces';

function getSourceFile(file: string, content: string) {
  const normalizedFile = path.normalize(file).replace(/\\/g, '/');
  return ts.createSourceFile(normalizedFile, content, ts.ScriptTarget.ES5, true);
}

export function _applyMods(content: string, mods: TypescriptMod[]) {
  let index = 0;
  let newContent: string[] = [];

  mods.forEach(mod => {
    newContent.push(content.slice(index, mod.start));
    newContent.push(mod.replacement);
    index = mod.start + mod.length;
  });

  newContent.push(content.slice(index));

  return newContent.join('');
}

export function _sortMods(mods: TypescriptMod[]) {
  const sorted = mods.sort((a, b) => {
    return a.start < b.start ? -1 : a.start === b.start ? 0 : 1;
  });

  // Check for overlapped ranges
  for (let i = 1; i < sorted.length; i++) {
    const prev = sorted[i - 1];
    const curr = sorted[i];
    if (prev.length + prev.start > curr.start) {
      throw new Error('Some of the mods overlap with each other. Please separate this out as another "asTypescript()" call.');
    }
  }

  return sorted;
}

export function _visitAndCollectMods(sourceFile: ts.SourceFile, visitor: Visitor) {
  const mods: TypescriptMod[] = [];

  const visitorWrapper = (node: ts.Node) => {
    const mod = visitor(node, modder);
    if (mod) {
      mods.push(mod);
    }
    node.forEachChild(visitorWrapper);
  };

  sourceFile.forEachChild(visitorWrapper);

  return mods;
}

export async function tsHandler(matches: string[], visitor: Visitor) {
  let sourceFile: ts.SourceFile;
  let content: string;
  let newContent: string;

  matches.forEach(async match => {
    content = fs.readFileSync(match).toString();

    try {
      sourceFile = getSourceFile(match, content);
    } catch {
      console.error('invalid Typescript file');
    }

    if (visitor) {
      const mods = _visitAndCollectMods(sourceFile, visitor);
      newContent = _applyMods(sourceFile.getFullText(), _sortMods(mods));

      if (newContent !== content) {
        fs.writeFileSync(match, newContent);
      }
    }
  });
}
