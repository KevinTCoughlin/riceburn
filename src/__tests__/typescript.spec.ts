import mockfs from 'mock-fs';
import fs from 'fs';
import ts from 'typescript';
import { tsHandler } from '../typescript';

const FixturesPath = 'src/__tests__/fixtures';

describe('tshandler', () => {
  const mockContents: { [pathName: string]: { [fileName: string]: string } } = {
    [FixturesPath]: {}
  };

  beforeAll(() => {
    console.log(process.cwd());

    const fixtures = fs.readdirSync(FixturesPath);

    fixtures.forEach(fixture => {
      if (fixture.endsWith('.ts') || fixture.endsWith('.tsx')) {
        mockContents[FixturesPath][fixture] = fs.readFileSync(`${FixturesPath}/${fixture}`).toString();
      }
    });
  });

  beforeEach(() => {
    mockfs(mockContents);
  });

  it('replaces node with text', () => {
    const fixture = `${FixturesPath}/variableDeclaration.ts`;

    tsHandler([fixture], (node, modder) => {
      if (ts.isVariableDeclaration(node)) {
        return modder.replace(node, `${node.name.getText()} = 'test'`);
      }
    });

    matchSnapshot(fs.readFileSync(fixture).toString());
  });

  it('prepends node with text', () => {
    const fixture = `${FixturesPath}/prepend.ts`;

    tsHandler([fixture], (node, modder) => {
      if (ts.isVariableDeclaration(node)) {
        return modder.prepend(node.parent, `console.log('prepended');\n`);
      }
    });

    matchSnapshot(fs.readFileSync(fixture).toString());
  });

  it('appends node with text', () => {
    const fixture = `${FixturesPath}/append.ts`;

    tsHandler([fixture], (node, modder) => {
      if (ts.isVariableDeclaration(node)) {
        return modder.append(node.parent, `\nconsole.log('appended');`);
      }
    });

    matchSnapshot(fs.readFileSync(fixture).toString());
  });

  it('removes node', () => {
    const fixture = `${FixturesPath}/remove.ts`;

    tsHandler([fixture], (node, modder) => {
      if (ts.isExpressionStatement(node)) {
        return modder.remove(node);
      }
    });

    matchSnapshot(fs.readFileSync(fixture).toString());
  });

  it('removes node fully', () => {
    const fixture = `${FixturesPath}/removeFull.ts`;

    tsHandler([fixture], (node, modder) => {
      if (ts.isExpressionStatement(node)) {
        return modder.removeFull(node);
      }
    });

    matchSnapshot(fs.readFileSync(fixture).toString());
  });

  function matchSnapshot(content: string) {
    mockfs.restore();
    expect(content).toMatchSnapshot();
  }
});
