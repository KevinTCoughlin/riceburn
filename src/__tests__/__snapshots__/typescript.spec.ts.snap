// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tshandler appends node with text 1`] = `
"// Here's a comment
var a = 1;

var b = 2;

var c = 3;
"
`;

exports[`tshandler prepends node with text 1`] = `
"// Here's a comment
var a = 1;

var b = 2;

var c = 3;
"
`;

exports[`tshandler removes node 1`] = `
"// Here's a comment
var a = 1;

var b = 2;

console.log('remove me');

var c = 3;
"
`;

exports[`tshandler removes node fully 1`] = `
"// Here's a comment
var a = 1;

var b = 2;

console.log('remove me fully');

var c = 3;
"
`;

exports[`tshandler replaces node with text 1`] = `
"// Here's a comment
var a = 1;

var b = 2;

var c = 3;

console.log(a);

console.log(b);
"
`;
